ISSUE-561 JMAP specs for calendar event parsing extension (#582)
linagora/tmail-backend: docs/modules/ROOT/pages/tmail-backend/jmap-extensions/calendarEventParsing.adoc

= Calendar event parse
:navtitle: calendar event parse

Allows users to parse calendar events for getting event information from the blob id of the calendar file (.ics).

== Additions to the capability object

Servers supporting the calendar event parse extension need
to advertise it through the session capabilities:
....
com:linagora:params:calendar:event
....

The associated object is empty.

== CalendarEvent/parse

> This method bases on:

Reviewer: ```suggestion
This method is based on:
```

- Parsing a Received MDN method as described in the RFC-9007

> - JMAP CalendarEvent RFC-8620

Reviewer: jmap calendar is still a draft (RFC8620 is the core spec), so wrong link here

-> https://datatracker.ietf.org/doc/html/draft-ietf-jmap-calendars

- iCalendar message RFC-5545

The request object requires the following properties:

- `accountId`: accountId of action user

> - `blobIds`: an array blobId of calendar which user wants to parse.

Reviewer: Can we present properties more in a jmap spec way?

```
- *accountId*: `Id` The id of the account to use.
- *blobIds*: `[Id]` An array of calendar blob ids that the user wants to parse.
```

Same for the response properties

The response object contains the following properties:

- `accountId`: accountId of action user
- `parsed`: a map of blobId and the corresponding calendar parse result.
- `notFound`: an array of blobId is not found.
- `notParsable`: an array of blobId has been found but is not parsable.

== Example

> The client's request to parse the calendar event which has blobId is `0f9f65ab-dc7b-4146-850f-6e4881093965`

Reviewer: ```suggestion
The client makes a request to parse the calendar event with a BlobId
```

            "duration": "PT2H0M0S",
            "end": "2023-02-09T12:00:00",
            "timeZone": "Asia/Ho_Chi_Minh",
            "location": "5 Dien Bien Phu, Ha Noi",
            "organizer": {
                "name": "Benoît TELLIER",
                "mailto": "btellier@linagora.com"
            },
            "participants": [{
                    "name": "Benoît TELLIER",
                    "mailto": "btellier@domain.tld",
                    "kind": "individual",
                    "role": "chair",
                    "participationStatus": "accepted",
                    "expectReply": false
                },
                {
                    "name": "Van Tung TRAN",
                    "mailto": "vttran@domain.tld",
                    "kind": "individual",
                    "role": "requested-participant",
                    "participationStatus": "needs-action",
                    "expectReply": true
                }
            ],
            "extensionFields": {
                "X-OPENPAAS-VIDEOCONFERENCE": "https://jitsi.linagora.com/abcd",
                "X-OPENPAAS-CUSTOM-HEADER1": "whatever"
            }
        }
    }
}, "c1" ]]
```

> The `parsed` value contains the following properties:

Reviewer: I believe the properties part of the parsed value should be explained iin the previous section where we define the object, and not along with the example

> - `title`: title of the event ~ SUMMARY (iCalendar)

Reviewer: We should precise that the properties match with the properties defined in rfc5545 (I mean SUMMARY, DURATION, DTSTART, etc). It's not necessary obvious reading it just like that

== CalendarEvent/parse

This method is based on:
- Parsing a Received MDN method as described in the RFC-9007
- Draft JMAP CalendarEvent (https://datatracker.ietf.org/doc/html/draft-ietf-jmap-calendars)
- iCalendar message RFC-5545

The request object requires the following properties:

- *accountId*: `Id` The id of the account to use.
- *blobIds*: `[Id]` An array of calendar blob ids that the user wants to parse.

The response object contains the following properties:

- *accountId*: `Id` The id of the account to use.
- *parsed*: a map of blob id and the corresponding calendar parse result.
- *notFound*: an array of blob id is not found.

> - *notParsable*: an array of blob id has been found but is not parsable.

Reviewer: ```
- *parsed*: `Map[BlobId]` and the corresponding calendar parse result.
- *notFound*: `[BlobId]` is not found.
- *notParsable*: `[BlobId]` has been found but is not parsable.
```

The `parsed` value contains the following properties (the properties below match with the properties defined in rfc5545):

> - `title`: title of the event

Reviewer: We don't have anymore the link with elements from iCalendar? SUMMARY, etc...
I didn't say to remove them, I said that we need a way for the reader to understand better where it is coming from :)

- `description`: description of the event
- `start`: Start time of the event
- `end`: End time of the event
- `duration`: duration of the event
- `timeZone`: time zone of the start/end time
- `location`: location of the event
- `organizer`: the organizer of the event, contains name (CN) and mail (mailto)
- `participants`: a collection of attendees

Each `participant` contains information:

- `name`: name of the participant
- `mailto`: mail address of the participant
- `kind`: to indicate the type of calendar user
- `role`: for the intended role that the attendee will have in the calendar component
- `participationStatus`: for the status of the attendee's participation
- `expectReply`: Boolean. for indicating whether the favor of a reply is requested

> - `extensionFields`: contains custom event headers.

Reviewer: Blank line missing
